UserDetailsClient/UserDetailsClient.Core/Features/LogOn/B2CAuthenticationService.cs

```diff
@@ -0,0 +1,157 @@
+﻿using Microsoft.Identity.Client;
+using Newtonsoft.Json.Linq;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace UserDetailsClient.Core.Features.LogOn
+{
+    public class B2CAuthenticationService : IAuthenticationService
+    {
+        public IPublicClientApplication PCA = null;
+
+        public object ParentActivityOrWindow { get; set; }
+
+        public B2CAuthenticationService()
+        {
+            // default redirectURI; each platform specific project will have to override it with its own
+            PCA = PublicClientApplicationBuilder.Create(B2CConstants.ClientID)
+                .WithB2CAuthority(B2CConstants.Authority)
+                .WithIosKeychainSecurityGroup("com.microsoft.adalcache")
+                .WithRedirectUri($"msal{B2CConstants.ClientID}://auth")
+                .Build();
+        }
+
+        public async Task<string> AcquireToken()
+        {
+            IEnumerable<IAccount> accounts = await PCA.GetAccountsAsync();
+            AuthenticationResult ar = await PCA.AcquireTokenSilent(B2CConstants.Scopes, GetAccountByPolicy(accounts, B2CConstants.PolicySignUpSignIn))
+                .WithB2CAuthority(B2CConstants.Authority)
+               .ExecuteAsync();
+            string token = ar.AccessToken;
+            return token;
+        }
+
+        public async Task<UserContext> ResetPassword()
+        {
+            AuthenticationResult ar = await PCA.AcquireTokenInteractive(B2CConstants.Scopes)
+                .WithPrompt(Prompt.NoPrompt)
+                .WithAuthority(B2CConstants.AuthorityPasswordReset)
+                .WithParentActivityOrWindow(ParentActivityOrWindow)
+                .ExecuteAsync();
+
+            var userContext = UpdateUserInfo(ar);
+
+            return userContext;
+        }
+
+        public async Task<UserContext> EditProfile()
+        {
+            IEnumerable<IAccount> accounts = await PCA.GetAccountsAsync();
+            // KNOWN ISSUE:
+            // User will get prompted 
+            // to pick an IdP again.
+            AuthenticationResult ar = await PCA.AcquireTokenInteractive(B2CConstants.Scopes)
+                .WithAccount(GetAccountByPolicy(accounts, B2CConstants.PolicyEditProfile))
+                .WithPrompt(Prompt.NoPrompt)
+                .WithAuthority(B2CConstants.AuthorityEditProfile)
+                .WithParentActivityOrWindow(ParentActivityOrWindow)
+                .ExecuteAsync();
+
+            var userContext = UpdateUserInfo(ar);
+
+            return userContext;
+        }
+
+        public async Task<UserContext> SignIn()
+        {
+            IEnumerable<IAccount> accounts = await PCA.GetAccountsAsync();
+            AuthenticationResult ar = await PCA.AcquireTokenInteractive(B2CConstants.Scopes)
```

Comment:
I understand what you are trying to do here but this isn't the pattern that we are typically trying to implement. The pattern we are trying to get ppl to use is: 

```csharp
try 
{
 // acquire token silent
}
catch (MsalUiRequiredException)
{
// acquire token interactive
}
```

So it should be simpler to get rid of AcquireToken method altogher and put the silent logic into SingIn.

Reply:
gotchya

Reply:
ok i fixed this...looking sweet...i like this much better...